### DISCUSSION TITLE
Best approach to SSE with Sveltekit between Pocketbase and client?

### DISCUSSION THREAD

**Discussion by bouchtaoui-dev** [pocketbase/pocketbase#3792] [Q&A]
Hi there,

I have Pocketbase as my backend, Sveltekit as the webserver and webclient.
So I have the following configuration:

Pocketbase <------> SvelteKit <------> Client

I want SSE between Pocketbase and client and the first thing I came up was this:

|--------------------------------------|
| Pocketbase <---SSE---> SvelteKit |<---SSE---> Client
|--------------------------------------|

I don't want Pocketbase be directly accessible.
Is this a good approach, or is there a better solution?

Thank you.

**Comment by ganigeorgiev** [comment]
> I don't want Pocketbase be directly accessible.

A lot of people seems to do this but I don't understand why. PocketBase is designed to be accessed directly.

What would be the benefit of proxying every request/message through the node server?

This may not be that important for your use case but note that if you proxy every client request through the node server, you are also limiting the PocketBase performance by the single threaded nature of node (_there are options for node workers/clustering and/or pm2 but that's another thing that you'll have to take care for_).

I'm not saying that JavaScript SSR is bad, but if you are not working on a website it comes with a lot of complications especially when interacting with 3rd party auth solutions for very little benefit (at least for me).

**Reply by einarpersson** [reply]
@ganigeorgiev 

I think it would be very helpful if there was a short mini-guide on how you can _combine_ Sveltekit server <->pocketbase with Sveltekit client <-> pocketbase. How do I get the best of both worlds, and when should I use which, and how do they work together?

Looking [here](https://github.com/pocketbase/js-sdk#ssr-integration) I can see info about how to use Pocketbase with Sveltekit SSR, but how can this approach be combined with client side? (either for enabling realtime subscriptions, or just out of simplicity wherever SSR is not needed).

I am using the SSR-approach (loading and exporting cookie in hooks file), but then when I wanted realtime functionality I added this to one of my pages:
```ts
const pb = new PocketBase(env.PUBLIC_POCKETBASE_URL) as TypedPocketBase

	onMount(async () => {
		pb.authStore.loadFromCookie(document.cookie, 'user')

		await pb.collection<UsersResponse<CourseProgress[]>>('users').subscribe('*', ({ record }) => {
			// ...
		})
	})
```

which works on a single +page.svelte, but I am unsure about what is a the recommended approach (should this be set in a store, context? can I add a global handler which manages what happens when the auth fails and redirects? what about universal loaders (+page.ts) files? etc)

Actually, if this works well I will probably replace a lot of sveltekit actions since it feels like I am just duplicating validation. But I would also like to [add typescript to pb_hooks](https://github.com/pocketbase/pocketbase/discussions/3341) in that case

**Reply by einarpersson** [reply]
I.e. I think that one answer to your question may be that they are just not sure how to do it in a safe and predictable way, and how you may combine these strategies if you want to start adding SSR to some parts

**Reply by ganigeorgiev** [reply]
I have it in my todo to convert the "Client-side SDKS" to "Client-side integrations" section in the docs and will consider elaborating with some examples.

But for now my personal preference and recommendation is to interact with PocketBase APIs directly and use client-side rendering (eg. SPA). The main benefit of doing this is that the regular user state is managed only between the browser/client and the PocketBase server.

If you need some extra server-side handling (eg. captcha validations, payment webhook verification, etc.) you could:

- [Use Pocketbase as framework and extend with Go/JS](https://pocketbase.io/docs/use-as-framework/).
- Create one-off node server-side actions that will interact with PocketBase only as admin and as pure data store (this is like regular database access but over HTTP).
  To initialize an admin client it is enough to authenticate as admin (in this case it is also safe to have a global JS SDK instance). For example:
  ```js
  // src/pb.js
  import PocketBase from "pocketbase"
 
  const pb = new PocketBase("https://example.com");

  pb.autoCancellation(false);

  // the credentials can be filled via env variables
  await pb.admins.authWithPassword(ADMIN_EMAIL, ADMIN_PASS, {
    // This will trigger auto refresh or auto reauthentication in case
    // the token has expired or is going to expire in the next 30 minutes.
    autoRefreshThreshold: 30 * 60
  })

  export default pb;
  ```
  Then you can just import the file in your server-side components and use it as usual:
  ```js
  import pb from "./src/pb.js"
  
  ... do some extra data validations or handling ...
  
  // send an "example" create request as admin
  await pb.collection("example").create({ ... })
  ```

**Reply by einarpersson** [reply]
@ganigeorgiev Thank you for rapid answers as always. I'll look at those examples!

As people starting with Svelte now typically do so by creating a Sveltekit app (and you may be completely right in that the SSR-stuff it imply often is overkill), this leads to learning about Sveltekit actions etc, and with it comes a creeping uncertainty about and perhaps a bias against direct client<->db communication. That is my hypothesis at least.

And also, as I said I think typescript is important to many and not having it in `pb_hooks` (by default) may lead to less people feeling comfortable with adding core logic there (and then there are the special rules related to the goja runtime).

Regarding admin: one issue with using admin is that it deactivates all validation / API-rules as to my knowledge. So then I once more have to split validation to both server and pocketbase, or add it in pb_hooks. I would prefer to be able to co-locate all "rules" to one place that governs what is allowed to get into the db, so that I don't create invalid states by accident.

---
To be clear: I think you have a good point. But I can also see where people are coming from, and that there still may be valid reasons why people want to _"just make sure everything goes through my server"_. 

I will probably go in the direction you describe, anyhow

**Comment by bouchtaoui-dev** [comment]
I end up with the following steps:

1. client authenticate to sveltekit, sveltekit authenticates to pocketbase
2. pocketbase returns authentication, sveltekit creates cookie with auth data (pb.authStore.exportToCookie()) and returns to client
3. sveltekit than clears pocketbase instance and user info to prevent spreading credentials to other client
4. since client received cookie with auth data, it uses to perform subsequent requests by passing along the cookie with auth data to sveltekit
5. sveltekit loads the client's cookie to perform requests to pocketbase on behalf of the client
6. client subscribe for SSE directly to pocketbas, as it has the cookie with auth data to create pocketbase instance to communicate directly with pocketbase
7. if a client has JS disabled, all requests is still possible via sveltekit, except SSE is not possible

This way I don't have mixed backends, I only have to deal with SvelteKit. One exception, only when I use SSE I have a direct communication with PocketBase.
The image below shows set up of my app.
![schema svelte pocketbase](https://github.com/pocketbase/pocketbase/assets/428598/59345440-5cb0-4502-89ad-2246730bbc94)

By the way, this works nice for me.

**Reply by MyraManes** [reply]
@bouchtaoui-dev, do you have an example repo of your setup that you could share?

**Reply by danfascia** [reply]
This is an interesting read but I do wish that @ganigeorgiev would come round to understanding a little more how people work rather than just denouncing what people do as a dumb idea... go use an SPA. Fundamentally I do not want client only routing and client side content protection, nor do I want to handle that only at API layer level and deal with the myriad of error states on the front end.

I've seen his opinion on "why would you use pocketbase as a backend for Astro for example" when you can fire up Drizzle, SQLite, Lucia, Resend, S3 storage, Litestream, some serverside functions for hooks etc by themselves... and therein lies the answer, because building out all of that is a shit ton of work and you still don't end up with the simple visual schema management and wonderful admin panel of Pocketbase. So please don't undersell your own fantastic creation.

Pocketbase offers a whole stack under one roof that is well coded, runs lean and fast on a low powered server. That's masive.

The use case that most people are aiming for (be it Sveltekit, Astro, Next, Nuxt whatever JS meta framework) goes like this:
- They want to authenticate (probably with OAuth, username, email + pw) and end up with a session cookie
- They want to have protected SSR routes to control access based on the validity of that cookie
- They need server side functions to act on webhooks (stripe etc)
- They want to leverage the lovely admin panel, schema building and JS SDK of pocketbase
- They want to make use of the realtime subscriptions for direct client side real time services (for likes, comments etc...)

So Pocketbase can potentially solve so much of this out of the box and the only thing that sits uneasy and difficult to solve is the decision to have no session management and be "stateless". Of course you can make use of the exportToCookie function but it still isn't tied to a session and relies on cookie expiry unless I can be informed otherwise.

Is it so bad that we all want to use Pocketbase in this way? and would it be such a ballache to just add conventional database stored sessions to the auth so that it can work either stateless like now or as a conventional authentication state server?

**Reply by ganigeorgiev** [reply]
@danfascia I nowhere said that this is a _"dumb idea"_ but in any case I'm not willing to discuss this further. Users are free to do whatever they want but as mentioned here and in several other discussions maintaining the auth state between 2 different backends comes with a lot of complications and I advice against it because I've stumbled on too many online tutorials and boilerplate repos that don't fully understand the context where their code is running leading to hard to debug security issues. If you know what you are doing then that doesn't apply for you.

**Reply by danfascia** [reply]
Appreciate the guidance - but perhaps it would be useful if you ever get the time to publish some posts / tutorials on "how to work with pocketbase" in popular frameworks or application models, and this may then save us wading into this territory. I know you didn't say it was a dumb idea but we inferred that from your response!

**Reply by gboujaoude** [reply]
@ganigeorgiev Thank you for all the work that you do! This is such a cool product, and I've told many people about it! Keep up the amazing work!
